From b3b412f01aa3913ff26c611950c48280a1f555fa Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Thu, 20 Sep 2018 02:47:16 +0800
Subject: [PATCH 4691/5242] MGS-4265 gpu: imx: dpu-blit: fix video hang with
 g2d compositor

commit  d284b041e53f7c3e7f505ae2e68c322fa37583b0 from
https://source.codeaurora.org/external/imx/linux-imx.git

video playback cause system hang with Wayland g2d compositor,
this also can be reproduced with Android G2D HWComposer.

the problem is second prg not handled between GPU and video.
need re-enable dprc & prg pipes when modifier changed.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu-blit/dpu-blit.c |   12 +++++++++++-
 drivers/gpu/imx/dpu-blit/dpu-blit.h |    2 ++
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit.c b/drivers/gpu/imx/dpu-blit/dpu-blit.c
index 4477890..65575db 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit.c
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit.c
@@ -108,6 +108,7 @@ void dpu_be_configure_prefetch(struct dpu_bliteng *dpu_be,
 			       u64 baddr, u64 uv_addr)
 {
 	struct dprc *dprc;
+	bool dprc_en=false;
 
 	/* Enable DPR, dprc1 is connected to plane0 */
 	dprc = dpu_be->dprc[1];
@@ -131,6 +132,13 @@ void dpu_be_configure_prefetch(struct dpu_bliteng *dpu_be,
 		return;
 	}
 
+	if (dpu_be->modifier != modifier && !dpu_be->start) {
+		dprc_disable(dprc);
+		dprc_en = true;
+	}
+
+	dpu_be->modifier = modifier;
+
 	dprc_configure(dprc, 0,
 		       width, height,
 		       x_offset, y_offset,
@@ -140,7 +148,7 @@ void dpu_be_configure_prefetch(struct dpu_bliteng *dpu_be,
 		       dpu_be->start,
 		       false);
 
-	if (dpu_be->start) {
+	if (dpu_be->start || dprc_en) {
 		dprc_enable(dprc);
 	}
 
@@ -415,6 +423,8 @@ int dpu_bliteng_init(struct dpu_bliteng *dpu_bliteng)
 	dpu_bliteng->start = true;
 	dpu_bliteng->sync = false;
 
+	dpu_bliteng->modifier = 0;
+
 	return 0;
 }
 EXPORT_SYMBOL_GPL(dpu_bliteng_init);
diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit.h b/drivers/gpu/imx/dpu-blit/dpu-blit.h
index fd6b500..75215b2 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit.h
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit.h
@@ -41,6 +41,8 @@ struct dpu_bliteng {
 
 	bool start;
 	bool sync;
+
+	u64 modifier;
 };
 
 #endif
-- 
1.7.9.5

